### PR TITLE
src: lib: git: Implement a git lib in src/lib/git.sh

### PR DESCRIPTION
**Pull request by BrizaMel**
Create lib that provides git easy functionalities for kworkflow code.

#1123

### CODE REVIEW

File: src/lib/git.sh
Comment (rodrigosiqueira) [inline src/lib/git.sh:1]: Please update the commit message to explain why we want to create this file and which features will be affected by this library. Additionally, add a reference to the commit message.

File: src/lib/git.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# The git.sh file provides git easy functionalities for kworkflow code.
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:1]: Just one overall comment for this file. Keep in mind that the bar for lib files should be as high as possible since many other files will use it, and it is easy to add tests. Remember, a library is only useful if it is reliable; try to be very rigorous about validations and corner cases here.

File: src/lib/git.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+# Run git bisect
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:3]: Please expand this documentation to explain how this function is expected to work and possible pitfalls that a developer may encounter.
P.S.: Review all your documentation for this file. Remember, this is a library file and we want to keep it easy to understand.

File: src/lib/git.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#
+# @good: good commit hash
+# @bad: bad commit hash
+# @command: test performed by git bisect
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:7]: Document returns

File: src/lib/git.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+function run_git_bisect()
+{
+  local good="$1"
+  local bad="$2"
+  local command="$3"
+
+  git bisect start
+  if [[ $? != 0 ]]; then
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:15]: Add double quotes around `$?`. Also, it is worth exploring if git bisect has different return errors that we can handle here.

File: src/lib/git.sh [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    printf '%s\n' 'git bisect start failed'
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:16]: Use the `complain` function.
The same comment is valid for this entire PR.

File: src/lib/git.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    exit 22 # EINVAL
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:17]: Avoid exit in a library function since this makes the debug very hard. Since this is a library, we should return an error code and let the caller handle the possible failure.

File: src/lib/git.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  fi
+  git bisect good "$good"
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:19]: Add a new line between these two lines.

File: src/lib/git.sh [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+# Run git bisect
+#
+# @good: good commit hash
+# @bad: bad commit hash
+# @command: test performed by git bisect
+function run_git_bisect()
+{
+  local good="$1"
+  local bad="$2"
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:11]: You can't trust the user input. Please validate it before using it.

File: src/lib/git.sh [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    exit 22 # EINVAL
+  fi
+  git bisect run "$command"
+  if [[ $? != 0 ]]; then
+    printf '%s\n' 'command run failed'
+    exit 22 # EINVAL
+  fi
+  return 0
+}
+
```
Comment (rodrigosiqueira) [inline src/lib/git.sh:36]: iirc, we have some validation for the SHA in the build function. That validation should be part of this lib.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi,
Thanks for this PR, I added some comments. However, when working on this feature, keep these things in mind:
1. You are writing a library component that must be useful in many different places. In this sense, you should be very picky about all your changes to ensure a good consistency for the lib.
2. Documentation is crucial for libs.
3. Check where we use the git command in the entire code from kw, and, based on that, try to abstract our usage of those commands as a library where it makes sense.
4. After you create the library, also add commits that replace the use of git in the kw source code with the new library functions.
5. Test is crucial here!

Thanks

**Comment by rodrigosiqueira** [comment]
@BrizaMel any update here?

**Comment by rodrigosiqueira** [comment]
Closed due to the lack of activity.
